### PR TITLE
bug fix in cyl mesh face inner products

### PR DESCRIPTION
**Pull request by lheagy**
- use sympy to generate a test showing the bug in the cyl mesh inner product
- more robust testing on cyl mesh
- bug fix in cyl mesh inner products

### CODE REVIEW

File: SimPEG/Mesh/CylMesh.py [hunk 1]
```diff
@@ -358,17 +358,15 @@ def _fastInnerProduct(self, projType, prop=None, invProp=False, invMat=False):
         if prop.size == self.nC:
             Av = getattr(self, 'ave'+projType+'2CC')
             Vprop = self.vol * Utils.mkvc(prop)
-            M = self.dim * Utils.sdiag(Av.T * Vprop)
+            nelts = np.sum(getattr(self, 'vn'+projType).nonzero())
+            M =  nelts * Utils.sdiag(Av.T * Vprop)
```
Comment (lheagy) [inline SimPEG/Mesh/CylMesh.py]: @rowanc1 : can you help me sort through this conceptually?

**Comment by lheagy** [comment]
@rowanc1, see what you think!

File: .travis.yml [hunk 2]
```diff
@@ -42,7 +42,7 @@ before_install:
   - conda update --yes conda
 
 install:
-  - conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython nose vtk sphinx
+  - conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython nose vtk sphinx sympy
```
Comment (lheagy) [inline .travis.yml]: are you good with sympy being a dependency? (makes writing those types of tests a bit easier)

File: SimPEG/Mesh/CylMesh.py [hunk 3]
```diff
@@ -383,46 +428,67 @@ def _fastInnerProductDeriv(self, projType, prop, invProp=False, invMat=False):
...
+                dMdprop = nelts * Utils.sdiag(- MI.diagonal()**2) * Av.T * V
 
         if tensorType == 1:
             Av = getattr(self, 'ave'+projType+'2CC')
             V = Utils.sdiag(self.vol)
+            nelts = np.sum(getattr(self, 'vn'+projType).nonzero())
+
             if not invMat and not invProp:
-                dMdprop = self.dim * Av.T * V
+                dMdprop = nelts * Av.T * V
```
Comment (lheagy) [inline SimPEG/Mesh/CylMesh.py]: is `nelts` (number of elements, 1 for edges, 2 for faces on cyl mesh) a clear enough variable name? any other suggestions

File: .travis.yml (hunk 2 above)
Comment (rowanc1) [inline .travis.yml]: We are not putting this in the req.txt?

Here it is completely fine!

Comment (lheagy) [inline .travis.yml]: I did put it in req.txt, but it is not necessary, could be removed and only put in .travis

File: requirements.txt [hunk 4]
```diff
@@ -3,3 +3,4 @@ scipy
 ipython
 matplotlib
 cython
+sympy
```
Comment (rowanc1) [inline requirements.txt]: I would prefer this not to be a direct dep? Just in Travis.

File: SimPEG/Mesh/CylMesh.py (hunk 3 above)
Comment (rowanc1) [inline SimPEG/Mesh/CylMesh.py]: `n_elements`?

Does this generalize to the innerproducts file directly?

Comment (lheagy) [inline SimPEG/Mesh/CylMesh.py]: I have generalized it now. There are a couple if statements, which isn't great, but overall most of the code is repeated, so nice to have it back in the tensor mesh.

**Pull request merged**